### PR TITLE
Added functionality for a miner to request a specific size coinbase t…

### PR DESCRIPTION
**Pull request by lonoami**

### CODE REVIEW

**Comment by ptschip** [comment]
This looks really interesting...haven't had a chance to look in detail but I think a couple of unit tests would be in order.

**Comment by lonoami** [comment]
Good point @ptschip Forgot about unit tests.
 For development I have a shell script.

File: src/miner.cpp [hunk 1]
```diff
@@ -121,15 +121,76 @@ uint64_t BlockAssembler::reserveBlockSize(const CScript &scriptPubKeyIn)
...
+    if (tx.vout.size() > 1)
+        LOGA("Discarding more than one transaction: %d.", tx.vout.size());
+
+    tx.vout.resize(1);
+    tx.vout[0].scriptPubKey.clear();
+    tx.vout[0].nValue = 0;
+
+    // Arg coinbaseSize 329 and 330 cause getminingcandidate() to generate 329+1, 330+1 size coinbase tx.
+    // By only accepting multiples of 4 the above is mitigated.
+    if (coinbaseSize % 4 != 0)
```
Comment (awemany) [inline src/miner.cpp]: Won't this be true for some negative values of `coinbaseSize` as well? I would suggest to use 0 for signalling "auto size" (and using an unsigned integer) instead of minus one as seems it would simplify this logic.

**Review by awemany — CHANGES_REQUESTED** [review]

File: src/miner.cpp (hunk 1 above)
Comment (lonoami) [reply src/miner.cpp]: Good point. I agree that  `MkFixedSizeCoinbaseTx(CMutableTransaction &tx, int64_t coinbaseSize)` would be better using an unsigned int. 
Imo <0 for signalling is good because it gives you a range of signals.
Lmk if you (others?) feel it is important to change it.

**Comment by gandrewstone** [comment]
We need the parameter to be passed by the getminingcandidate RPC call as suggested by jtoomim here: https://github.com/btccom/btcpool/issues/89.

Additionally, there is no need to actually create a coinbase transaction of the specified size.  If 50k bytes are reserved for the coinbase and our coinbase is only 300 bytes then the block will be smaller.  Presumably the miner will throw away our coinbase and replace it with his own.

Take a look at miner.cpp:

    uint64_t BlockAssembler::reserveBlockSize(const CScript &scriptPubKeyIn)
    {
    CBlockHeader h;
    uint64_t nHeaderSize, nCoinbaseSize;

    // BU add the proper block size quantity to the actual size
    nHeaderSize = ::GetSerializeSize(h, SER_NETWORK, PROTOCOL_VERSION);
    assert(nHeaderSize == 80); // BU always 80 bytes
    nHeaderSize += 5; // tx count varint - 5 bytes is enough for 4 billion txs; 3 bytes for 65535 txs

    // This serializes with output value, a fixed-length 8 byte field, of zero and height, a serialized CScript
    // signed integer taking up 4 bytes for heights 32768-8388607 (around the year 2167) after which it will use 5.
    nCoinbaseSize = ::GetSerializeSize(coinbaseTx(scriptPubKeyIn, 400000, 0), SER_NETWORK, PROTOCOL_VERSION);

    // BU Miners take the block we give them, wipe away our coinbase and add their own.
    // So if their reserve choice is bigger then our coinbase then use that.
    return nHeaderSize + std::max(nCoinbaseSize, coinbaseReserve.Value());
    }

Basically if the miner passes a parameter into getminingcandidate we need it to replace coinbaseReserve.Value() in the last line of this function.  (but we don't want to change coinbaseReserve since that would make his choice persistent)

**Comment by lonoami** [comment]
I'll do what you suggest.

BTW The coinbasesize argument is implemented:
Miner:
../bu/src/bitcoin-miner -coinbasesize=300

bitcoind:

`UniValue getminingcandidate(const UniValue &params, bool fHelp)`
`mkblocktemplate(UniValue(UniValue::VARR), coinbaseSize, &candid.block);`
...
`uint64_t BlockAssembler::reserveBlockSize(const CScript &scriptPubKeyIn, int64_t coinbaseSize)`

File: src/unlimited.cpp [hunk 2]
```diff
@@ -1683,6 +1683,7 @@ static UniValue MkMiningCandidateJson(CMiningCandidate &candid)
     {
         const CTransaction *tran = block.vtx[0].get();
         ret.push_back(Pair("coinbase", EncodeHexTx(*tran)));
+        // printf("MkMiningCandidateJson coinbase %lu\n",EncodeHexTx(*tran).length());
```
Comment (gandrewstone) [inline src/unlimited.cpp]: delete this commented printf

**Comment by gandrewstone** [comment]
Please extend the miningtest.py QA test to exercise this new parameter.  And also try sending some illegal values and check for the proper error.

**Comment by lonoami** [comment]
BTW. The minimum size `UniValue getminingcandidate(const UniValue &params, bool fHelp)` accepts is 0 (zero). This probably does not make sense. 0 to 112 generate  the same size block (112+85). Any ideas as to what minimum size should be? Is there a constant I can use?

**Comment by gandrewstone** [comment]
Its fine.  Any size passed that is lower than our coinbase tx gets effectively rounded up to fit our coinbase tx.  We don't need the param to also enforce this minimum -- that way if we ever make our default coinbase tx smaller we won't have change unrelated code.
